node/pkg/watchers/evm/blocks_by_timestamp.go

```diff
@@ -0,0 +1,149 @@
+	Blocks []Block
+
+	Block struct {
+		Timestamp uint64
+		BlockNum  uint64
+	}
+)
+
+// NewBlocksByTimestamp creates an empty cache of blocks by timestamp.
+func NewBlocksByTimestamp(maxCacheSize int) *BlocksByTimestamp {
+	return &BlocksByTimestamp{
+		cache:        Blocks{},
+		maxCacheSize: maxCacheSize,
+	}
+}
+
+// AddLatest adds a block to the end of the cache. This is meant to be used in the normal scenario when a new latest block is received. It will return an error if the block should not go at the end.
+func (bts *BlocksByTimestamp) AddLatest(timestamp uint64, blockNum uint64) error {
+	bts.mutex.Lock()
+	defer bts.mutex.Unlock()
+	if len(bts.cache) > 0 {
+		last := &bts.cache[len(bts.cache)-1]
+		if timestamp < last.Timestamp {
+			return fmt.Errorf("timestamp %d must be greater than or equal to the previous latest %d", timestamp, last.Timestamp)
+		}
+		if blockNum <= last.BlockNum {
+			return fmt.Errorf("block number %d must be greater than the previous latest %d", blockNum, last.BlockNum)
+		}
+	}
+
+	bts.cache = append(bts.cache, Block{Timestamp: timestamp, BlockNum: blockNum})
+
+	if len(bts.cache) > bts.maxCacheSize {
+		bts.cache = bts.cache[1:]
+	}
+	return nil
+}
+
+// AddBatch adds a batch of blocks to the cache. This is meant to be used for backfilling the cache. It makes sure there are no duplicate blocks and regenerates the cache in the correct order by timestamp.
+func (bts *BlocksByTimestamp) AddBatch(blocks Blocks) {
+	bts.mutex.Lock()
+	defer bts.mutex.Unlock()
+
+	// First build a map of all the existing blocks so we can avoid duplicates.
+	blockMap := make(map[uint64]uint64)
+	for _, block := range bts.cache {
+		blockMap[block.BlockNum] = block.Timestamp
+	}
+
+	// Now add the new blocks to the map, overwriting any duplicates. (Maybe there was a reorg. . .)
+	for _, block := range blocks {
+		blockMap[block.BlockNum] = block.Timestamp
+	}
+
+	// Now put everything into the cache in random order.
+	cache := Blocks{}
+	for blockNum, timestamp := range blockMap {
+		cache = append(cache, Block{Timestamp: timestamp, BlockNum: blockNum})
+	}
+
+	// Sort the cache into timestamp order.
+	sort.SliceStable(cache, func(i, j int) bool {
+		return cache[i].Cmp(cache[j]) < 0
+	})
+
+	if len(cache) > bts.maxCacheSize {
+		// Trim the cache.
+		trimIdx := len(cache) - bts.maxCacheSize
+		cache = cache[trimIdx:]
+	}
+
+	bts.cache = cache
+}
+
+// LookUp searches the cache for the specified timestamp and returns the blocks surrounding that timestamp. It also returns true if the results are complete or false if they are not.
+// The following rules apply:
+// - If timestamp is less than the first timestamp in the cache, it returns (0, <theFirstBlockInTheCache>, false)
+// - If timestamp is greater than or equal to the last timestamp in the cache, it returns (<theLastBlockInTheCache>, 0, false)
+// - If timestamp exactly matches one in the cache, it returns (<theLastBlockForThatTimestamp>, <theFirstBlockForTheNextTimestamp>, true)
```

Comment:
it seems like this is what it returns if it matches one or more in the cache, right?

Reply:
A timestamp can only match one timestamp. :grin: